ksql-engine/src/test/java/io/confluent/ksql/util/ExecutorUtilTest.java

```diff
@@ -0,0 +1,132 @@
+    expectedException.expectMessage("I will never succeed");
+
+    ExecutorUtil.executeWithRetries(() -> {
+          final CompletableFuture<Void> f = new CompletableFuture<>();
+          f.completeExceptionally(new TestRetriableException("I will never succeed"));
+          return f.get();
+        },
+        ExecutorUtil.RetryBehaviour.ON_RETRYABLE);
+  }
+
+  @Test
+  public void shouldRetryAndSucceed() throws Exception {
+    final AtomicInteger counts = new AtomicInteger(5);
+
+    ExecutorUtil.executeWithRetries(() -> {
+      if (counts.decrementAndGet() == 0) {
+        return CompletableFuture.completedFuture(null).get();
+      }
+
+      final CompletableFuture<Void> f = new CompletableFuture<>();
+      f.completeExceptionally(new TestRetriableException("I will never succeed"));
+      return f.get();
+    },
+    ExecutorUtil.RetryBehaviour.ON_RETRYABLE);
+  }
+
+  @Test
+  public void shouldReturnValue() throws Exception {
+    final String expectedValue = "should return this";
+
+    assertThat(ExecutorUtil.executeWithRetries(
+        () -> CompletableFuture.completedFuture(expectedValue).get(),
+        ExecutorUtil.RetryBehaviour.ON_RETRYABLE),
+        is(expectedValue));
+  }
+
+  @Test
+  public void shouldNotRetryOnNonRetryableException() throws Exception {
+    expectedException.expect(RuntimeException.class);
+    expectedException.expectMessage("First non-retry exception");
+
+    final AtomicBoolean firstCall = new AtomicBoolean(true);
+
+    ExecutorUtil.executeWithRetries(() -> {
+      final CompletableFuture<Void> f = new CompletableFuture<>();
+
+      if (firstCall.get()) {
+        firstCall.set(false);
+        f.completeExceptionally(new RuntimeException("First non-retry exception"));
+      } else {
+        f.completeExceptionally(new RuntimeException("Test should not retry"));
+      }
+
+      return f.get();
+    }, ExecutorUtil.RetryBehaviour.ON_RETRYABLE);
+  }
+
+  @Test
+  public void shouldNotRetryIfSupplierThrowsNonRetryableException() throws Exception {
+    expectedException.expect(RuntimeException.class);
+    expectedException.expectMessage("First non-retry exception");
+
+    final AtomicBoolean firstCall = new AtomicBoolean(true);
+
+    ExecutorUtil.executeWithRetries(() -> {
+      if (firstCall.get()) {
+        firstCall.set(false);
+        throw new RuntimeException("First non-retry exception");
+      }
+
+      throw new RuntimeException("Test should not retry");
+    }, ExecutorUtil.RetryBehaviour.ON_RETRYABLE);
+  }
+
+  @Test
+  public void shouldRetryIfSupplierThrowsRetryableException() throws Exception {
+    final AtomicInteger counts = new AtomicInteger(5);
+
+    ExecutorUtil.executeWithRetries(() -> {
+      if (counts.decrementAndGet() == 0) {
+        return CompletableFuture.completedFuture(null).get();
+      }
+
+      throw new TestRetriableException("Test should retry");
+    }, ExecutorUtil.RetryBehaviour.ON_RETRYABLE);
+  }
+
```

Comment:
Could do with a test case that covers the case where the supplier throws an ExecutionException, which is wrapping a RetriableExeception`.